macrosynergy/management/utils/df_utils.py

```diff
@@ -901,3 +960,91 @@ def get_eops(
     t_dates: pd.Series = dts["real_date"].loc[t_indices].reset_index(drop=True)
 
     return t_dates
+
+
+def estimate_release_frequency(
+    timeseries: pd.Series = None,
+    df_wide: pd.DataFrame = None,
+    exception_tolerance: float = 0.1,
+    *args,
+    **kwargs,
+) -> Union[str, pd.DataFrame]:
+    """
+    Estimates the release frequency of a timeseries.
+
+    :param <pd.Series> timeseries: The timeseries to be used to estimate the release
+        frequency.
+    :param <float> exception_tolerance: The tolerance for exceptions in the release
+        frequency estimation. Must be a float between 0 and 1.
+
+    :return <str>: The estimated release frequency.
+
+    """
+
+    if df_wide is not None:
+        if timeseries is not None:
+            raise ValueError("Only one of `timeseries` or `df_wide` must be passed.")
+        if not isinstance(df_wide, pd.DataFrame):
+            raise TypeError("Argument `df_wide` must be a pandas DataFrame.")
+        if df_wide.empty or df_wide.index.name != "real_date":
+            raise ValueError(
+                "Argument `df_wide` must be a non-empty pandas DataFrame with a datetime "
+                "index `'real_date'`."
+            )
+
+        return _estimate_release_frequency_wide(df_wide=df_wide, *args, **kwargs)
+
+    timeseries: pd.Series = timeseries.copy().dropna().drop_duplicates(keep="first")
+
+    if (
+        not isinstance(timeseries, pd.Series)
+        or timeseries.empty
+        or not isinstance(timeseries.index, pd.DatetimeIndex)
+    ):
+        raise TypeError(
+            "Argument `timeseries` must be a non-empty pandas Series with "
+            "a datetime index `'real_date'`."
+        )
+
+    if not isinstance(exception_tolerance, (int, float)) or not (
+        0 <= exception_tolerance <= 1
+    ):
+        raise TypeError(
+            "Argument `exception_tolerance` must be a float between 0 and 1."
+        )
+
+    eops_scores = {}
+    for freqx in FREQUENCY_MAP.keys():
+        eops_scores[freqx] = (
+            timeseries.groupby(pd.Grouper(timeseries, freq="B")).nunique().mean()
+            / DAYS_PER_FREQ[freqx]
+        )
+
+    eops_df = pd.Series(eops_scores, name="score")
+
+    return eops_df.idxmax()
+
+
+def _estimate_release_frequency_wide(df_wide: pd.DataFrame, *args, **kwargs):
+    """
+    Estimates the release frequency of a wide DataFrame.
+    Backend for `estimate_release_frequency`.
+
+    :param <pd.DataFrame> df_wide: The wide DataFrame to be used to estimate the release
+        frequency.
+
+    :return <pd.DataFrame>: The estimated release frequency.
+    """
+    assert "timeseries" not in kwargs, "Argument `timeseries` is not allowed."
+    assert isinstance(df_wide, pd.DataFrame), "Argument `df_wide` must be a DataFrame."
+
+    return pd.Series(
+        [
+            estimate_release_frequency(
```

Comment:
How come the hidden function is calling the non-hidden function? Isn't it usually the other way around?

Reply:
The hidden function in this case exists specifically to make the main function more readable and to provide clean compatibility for wide dataframes and single timeseries (`pd.Series`). 
This is very similar to cases where we use:
```python
def fmt(x: str):
    if isinstance(inputX, list):
        return list(map(fmt, inputX))
    return x.capitalize()
```
where the hidden function can be thought of as a fancy `map`